MusicNotationKit/MusicNotationKit/Measure.swift

```diff
@@ -53,50 +55,327 @@ public struct Measure: ImmutableMeasure, Equatable {
-        notes[setIndex].append(note)
-        noteCount[setIndex] += note.noteCount
-    }
+    internal mutating func replaceNote(at collectionIndex: NoteCollectionIndex, with noteCollections: [NoteCollection], inSet setIndex: Int) throws {
+        guard collectionIndex.tupletIndex == nil || collectionIndex.tupletIndex == 0 else {
+            throw MeasureError.invalidTupletIndex
+        }
 
-    public mutating func insertNote(_ note: Note, at index: Int, inSet setIndex: Int = 0) throws {
-        // TODO: Implement
+        for note in noteCollections.reversed() {
+            try replaceNote(at: collectionIndex, with: note, inSet: setIndex)
+        }
     }
 
-    public mutating func removeNote(at index: Int, inSet setIndex: Int = 0) throws {
-        // TODO: Implement
+    internal mutating func removeNote(at index: Int, inSet setIndex: Int, skipTies skipTieConfig: Bool) throws {
+        var newMeasure = self
+
+        if !skipTieConfig {
+            try newMeasure.prepTiesForRemoval(at: index, inSet: setIndex)
+        }
+
+        let collectionIndex = try newMeasure.noteCollectionIndex(fromNoteIndex: index, inSet: setIndex)
+        guard collectionIndex.tupletIndex == nil else {
+            throw MeasureError.removeNoteFromTuplet
+        }
+        newMeasure.notes[setIndex].remove(at: collectionIndex.noteIndex)
+        self = newMeasure
     }
 
-    public mutating func removeNotesInRange(_ indexRange: Range<Int>, inSet setIndex: Int = 0) throws {
-        // TODO: Implement
+    internal mutating func removeNotesInRange(_ indexRange: CountableClosedRange<Int>, inSet setIndex: Int = 0, skipTies skipTieConfig: Bool) throws {
+        var newMeasure = self
+
+        if !skipTieConfig {
+            try newMeasure.prepTiesForRemoval(at: indexRange.lowerBound, inSet: setIndex)
+            try newMeasure.prepTiesForRemoval(at: indexRange.upperBound, inSet: setIndex)
+        }
+
+        for index in indexRange.reversed() {
+            let collectionIndex = try newMeasure.noteCollectionIndex(fromNoteIndex: index, inSet: setIndex)
+            if collectionIndex.tupletIndex != nil {
+                guard let tupletIndex = collectionIndex.tupletIndex,
+                    let tuplet = newMeasure.notes[setIndex][collectionIndex.noteIndex]  as? Tuplet else {
+                        assertionFailure("note collection should be tuplet, but cast failed")
+                        throw MeasureError.internalError
+                }
+                // Range starts with  an incomplete lower bound
+                if index == indexRange.lowerBound {
+                    guard tupletIndex == 0 else {
+                        throw MeasureError.incompleteTuplet
+                    }
+                }
+                // Error if provided indexRange does not cover the tuple.
+                guard tuplet.noteCount - tupletIndex - 1 <= indexRange.upperBound - index else {
+                    throw MeasureError.incompleteTuplet
```

Comment:
Think of a little more descriptive name, like `tupletNotCompletelyCovered` or something.